File: src/main/java/dev/raphaeldelio/redisapollo/service/TOCService.java
package dev.raphaeldelio.redisapollo.service;

import com.redis.om.spring.search.stream.EntityStream;
import dev.raphaeldelio.redisapollo.document.domain.TOCData;
import dev.raphaeldelio.redisapollo.document.domain.TOCData$;
import dev.raphaeldelio.redisapollo.hash.domain.Utterance;
import dev.raphaeldelio.redisapollo.hash.domain.Utterance$;
import dev.raphaeldelio.redisapollo.document.repository.TOCDataRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.ai.chat.messages.SystemMessage;
import org.springframework.ai.chat.messages.UserMessage;
import org.springframework.ai.chat.model.ChatResponse;
import org.springframework.ai.chat.prompt.Prompt;
import org.springframework.ai.openai.OpenAiChatModel;
import org.springframework.stereotype.Service;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

@Service
public class TOCService {

    private static final Logger logger = LoggerFactory.getLogger(TOCService.class);

    private final TOCDataRepository tocDataRepository;
    private final OpenAiChatModel chatModel;
    private final EntityStream entityStream;
    private final FileService fileService;
    private final RedisService redisService;

    public TOCService(TOCDataRepository tocDataRepository, OpenAiChatModel chatModel, EntityStream entityStream, FileService fileService, RedisService redisService) {
        this.tocDataRepository = tocDataRepository;
        this.chatModel = chatModel;
        this.entityStream = entityStream;
        this.fileService = fileService;
        this.redisService = redisService;
    }

    public void loadTOCData(String filePath) {
        logger.info("Loading TOC data from file: {}", filePath);
        fileService.readAndProcessFile(filePath, TOCData.class, data -> {
            for (TOCData tocData : data) {
                if (!isValidToc(tocData)) {
                    continue;
                }
                int utteranceSeconds = fileService.asSeconds(tocData.getStartDate());
                tocData.setStartDate(tocData.getStartDate().replace(":", ";"));
                tocData.setStartDateInt(utteranceSeconds);
                tocDataRepository.save(tocData);
            }
        });
    }

    private boolean isValidToc(TOCData tocData) {
        return !tocData.getDescription().startsWith("Video: ");
    }

    public void populateUtterances() {
        populateUtterances(false);
    }

    public void populateUtterances(boolean overwrite) {
        logger.info("Grouping utterances by TOC entries");

        List<TOCData> tocDataList = entityStream.of(TOCData.class)
                .sorted(TOCData$.START_DATE_INT)
                .collect(Collectors.toList());

        List<String> utterancesToFilterOut = redisService.getZRangeByScore("utterances", 2, Double.MAX_VALUE);

        // Iterate through each TOC entry
        for (int i = 0; i < tocDataList.size(); i++) {
            TOCData currentTOC = tocDataList.get(i);
            logger.info("Processing TOC entry: {}", currentTOC.getStartDateInt());

            if (!overwrite && currentTOC.getConcatenatedUtterances() != null && !currentTOC.getConcatenatedUtterances().isBlank()) {
                logger.info("Utterances already grouped for TOC entry: {}", currentTOC.getStartDate());
                continue;
            }

            int startDate = currentTOC.getStartDateInt();
            int endDate = (i < tocDataList.size() - 1) ? tocDataList.get(i + 1).getStartDateInt() : Integer.MAX_VALUE;

            // Filter utterances that fall within the current TOC's time range
            //utterancesToFilterOut.contains(utt.getText())
            List<Utterance> tocUtterances = entityStream.of(Utterance.class)
                    .filter(Utterance$.TIMESTAMP_INT.ge(startDate)
                            .and(Utterance$.TIMESTAMP_INT.le(endDate))
                    ).collect(Collectors.toList());

            tocUtterances = tocUtterances.stream().filter(utt ->
                    !utterancesToFilterOut.contains(utt.getText())
            ).toList();

            var groupedUtterances = tocUtterances.stream()
                    .map(utt -> utt.getSpeaker() + ":" + utt.getText())
                    .reduce((a, b) -> a + "\n" + b)
                    .orElse(null);

            if (groupedUtterances != null) {
                currentTOC.setConcatenatedUtterances(groupedUtterances);
                currentTOC.setUtterances(tocUtterances);
                tocDataRepository.save(currentTOC);
                logger.info("Grouped utterances for TOC entry: {}", currentTOC.getStartDate());
            } else {
                logger.info("No utterances found for TOC entry: {}", currentTOC.getStartDate());
            }
        }
        logger.info("Grouping complete");
    }

    public void summarize() {
        summarize(false);
    }

    public void summarize(boolean overwrite) {
        logger.info("Summarizing TOC entries");
        List<TOCData> tocDataList = tocDataRepository.findAll();

        // Filter TOC entries that need summaries generated
        List<TOCData> tocDataToProcess = tocDataList.stream()
                .filter(tocData -> {
                    boolean isUtterancesPopulated = tocData.getConcatenatedUtterances() != null && !tocData.getConcatenatedUtterances().isBlank();
                    boolean shouldOverwrite = overwrite || tocData.getSummary() == null;
                    return isUtterancesPopulated && shouldOverwrite;
                })
                .collect(Collectors.toList());

        logger.info("Found {} TOC entries that need summaries generated", tocDataToProcess.size());

        if (tocDataToProcess.isEmpty()) {
            logger.info("No TOC entries need summaries generated");
            return;
        }

        // Process in batches of 300
        int batchSize = 300;
        int totalBatches = (tocDataToProcess.size() + batchSize - 1) / batchSize; // Ceiling division

        logger.info("Processing TOC entries in {} batches of up to {} entries each", totalBatches, batchSize);

        // Create a virtual thread per task executor for better scalability with I/O-bound operations
        // Virtual threads are lightweight and don't consume OS resources like platform threads
        // This is ideal for I/O-bound tasks like API calls to OpenAI
        try (ExecutorService executorService = Executors.newVirtualThreadPerTaskExecutor()) {

            for (int batchIndex = 0; batchIndex < totalBatches; batchIndex++) {
                int startIndex = batchIndex * batchSize;
                int endIndex = Math.min(startIndex + batchSize, tocDataToProcess.size());
                List<TOCData> currentBatch = tocDataToProcess.subList(startIndex, endIndex);

                logger.info("Processing batch {} of {} ({} entries)", batchIndex + 1, totalBatches, currentBatch.size());

                // Process each batch asynchronously
                List<CompletableFuture<TOCData>> futures = currentBatch.stream()
                        .map(tocData -> CompletableFuture.supplyAsync(() -> {
                            try {
                                logger.info("Generating summary for TOC entry: {}", tocData.getStartDate());

                                ChatResponse response = chatModel.call(
                                        new Prompt(List.of(
                                                new SystemMessage("""
                                                        You are a helpful assistant who summarizes utterances of the Apollo 11 mission.
                                                        Make these summaries very dense with all curiosities included.
                                                        Limit the summary to 512 words.
                                                        """),
                                                new UserMessage(tocData.getConcatenatedUtterances())
                                        ))
                                );

                                tocData.setSummary(response.getResult().getOutput().getText());
                                logger.info("Successfully generated summary for TOC entry: {}", tocData.getStartDate());
                                return tocData;
                            } catch (Exception e) {
                                logger.error("Error generating summary for TOC entry: {}", tocData.getStartDate(), e);
                                return null;
                            }
                        }, executorService))
                        .toList();

                // Wait for all futures in the current batch to complete before processing the next batch
                CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();

                // Collect all processed TOCData objects and filter out nulls (from failed futures)
                List<TOCData> processedBatch = futures.stream()
                        .map(CompletableFuture::join)
                        .filter(Objects::nonNull)
                        .toList();

                // Save all processed TOCData objects in batch
                if (!processedBatch.isEmpty()) {
                    tocDataRepository.saveAll(processedBatch);
                    logger.info("Saved {} TOC entries in batch", processedBatch.size());
                }

                logger.info("Completed batch {} of {}", batchIndex + 1, totalBatches);
            }
        }

        logger.info("Completed generating summaries for all TOC entries");
    }

    public void generateQuestions() {
        generateQuestions(false);
    }

    public void generateQuestions(boolean overwrite) {
        logger.info("Generating questions for TOC entries");
        List<TOCData> tocDataList = tocDataRepository.findAll();

        // Filter TOC entries that need questions generated
        List<TOCData> tocDataToProcess = tocDataList.stream()
                .filter(tocData -> {
                    boolean isUtterancesPopulated = tocData.getConcatenatedUtterances() != null && !tocData.getConcatenatedUtterances().isBlank();
                    boolean shouldOverwrite = overwrite || tocData.getQuestions() == null;
                    return isUtterancesPopulated && shouldOverwrite;
                })
                .collect(Collectors.toList());

        logger.info("Found {} TOC entries that need questions generated", tocDataToProcess.size());

        if (tocDataToProcess.isEmpty()) {
            logger.info("No TOC entries need questions generated");
            return;
        }

        // Process in batches of 300
        int batchSize = 300;
        int totalBatches = (tocDataToProcess.size() + batchSize - 1) / batchSize; // Ceiling division

        logger.info("Processing TOC entries in {} batches of up to {} entries each", totalBatches, batchSize);

        // Create a virtual thread per task executor for better scalability with I/O-bound operations
        // Virtual threads are lightweight and don't consume OS resources like platform threads
        // This is ideal for I/O-bound tasks like API calls to OpenAI
        try (ExecutorService executorService = Executors.newVirtualThreadPerTaskExecutor()) {

            for (int batchIndex = 0; batchIndex < totalBatches; batchIndex++) {
                int startIndex = batchIndex * batchSize;
                int endIndex = Math.min(startIndex + batchSize, tocDataToProcess.size());
                List<TOCData> currentBatch = tocDataToProcess.subList(startIndex, endIndex);

                logger.info("Processing batch {} of {} ({} entries)", batchIndex + 1, totalBatches, currentBatch.size());

                // Process each batch asynchronously
                List<CompletableFuture<TOCData>> futures = currentBatch.stream()
                        .map(tocData -> CompletableFuture.supplyAsync(() -> {
                            try {
                                logger.info("Generating questions for TOC entry: {}", tocData.getStartDate());

                                ChatResponse response = chatModel.call(
                                        new Prompt(List.of(
                                                new SystemMessage("""
                                                        You are a helpful assistant that is helping me predict which questions can be asked by people who are trying to
                                                        rediscover the Apollo 11 mission data. You will be given a number of utterances and you will predict the questions that
                                                        can be asked by people who are trying to rediscover the Apollo 11 mission data. You will ONLY return the questions separate by breaklines,
                                                        and nothing more. You will NEVER return more than 512 words.
                                                        """),
                                                new UserMessage(tocData.getConcatenatedUtterances())
                                        ))
                                );

                                var questions = Arrays.stream(response.getResult().getOutput().getText().split("\n"))
                                        .filter(q -> !q.isBlank())
                                        .toList();

                                tocData.setQuestions(questions);
                                logger.info("Successfully generated {} questions for TOC entry: {}", questions.size(), tocData.getStartDate());
                                return tocData;
                            } catch (Exception e) {
                                logger.error("Error generating questions for TOC entry: {}", tocData.getStartDate(), e);
                                return null;
                            }
                        }, executorService))
                        .toList();

                // Wait for all futures in the current batch to complete before processing the next batch
                CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();

                // Collect all processed TOCData objects and filter out nulls (from failed futures)
                List<TOCData> processedBatch = futures.stream()
                        .map(CompletableFuture::join)
                        .filter(Objects::nonNull)
                        .toList();

                // Save all processed TOCData objects in batch
                if (!processedBatch.isEmpty()) {
                    tocDataRepository.saveAll(processedBatch);
                    logger.info("Saved {} TOC entries in batch", processedBatch.size());
                }

                logger.info("Completed batch {} of {}", batchIndex + 1, totalBatches);
            }
        }

        logger.info("Completed generating questions for all TOC entries");
    }
}

File: redis-apollo-app-kotlin/src/main/kotlin/dev/raphaeldelio/redisapollo/tableofcontents/TOCService.kt
package dev.raphaeldelio.redisapollo.tableofcontents

import com.redis.om.spring.search.stream.EntityStream
import dev.raphaeldelio.redisapollo.dataloader.FileService
import dev.raphaeldelio.redisapollo.RedisService
import dev.raphaeldelio.redisapollo.utterance.Utterance
import dev.raphaeldelio.redisapollo.utterance.`Utterance$`.TIMESTAMP_INT
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.stream.Collectors

@Service
class TOCService(
    private val tocDataRepository: TOCDataRepository,
    private val entityStream: EntityStream,
    private val fileService: FileService,
    private val redisService: RedisService
) {
    private val logger = LoggerFactory.getLogger(TOCService::class.java)

    fun loadTOCData(filePath: String) {
        logger.info("Loading TOC data from file: {}", filePath)
        fileService.readAndProcessFile(filePath, TOCData::class.java) { data ->
            data.filter(::isValidToc)
                .onEach { tocData ->
                    val startInt = fileService.asSeconds(tocData.startDate)
                    tocData.startDate = tocData.startDate.replace(":", ";")
                    tocData.startDateInt = startInt }
                .chunked(1000) { batch ->
                    tocDataRepository.saveAll(batch)
                }
        }
    }

    private fun isValidToc(tocData: TOCData): Boolean {
        return !tocData.description.startsWith("Video: ")
    }

    fun populateUtterances(overwrite: Boolean = false) {
        logger.info("Grouping utterances by TOC entries")
        val tocList = entityStream.of(TOCData::class.java)
            .sorted(`TOCData$`.START_DATE_INT)
            .collect(Collectors.toList())

        val toFilterOut = redisService.getZRangeByScore("utterances", 2.0, Double.MAX_VALUE)

        tocList.onEach { logger.info("Processing TOC entry: {}", it.startDateInt) }
            .filter { overwrite || it.concatenatedUtterances.isNullOrBlank() }
            .onEach {
                if (!overwrite) {
                    logger.info("Utterances already grouped for TOC entry: {}", it.startDate)
                }
            }
            .mapNotNull { toc ->
                val start = toc.startDateInt
                val end = tocList.getOrNull(tocList.indexOf(toc) + 1)?.startDateInt ?: Int.MAX_VALUE

                val utterances = entityStream.of(Utterance::class.java)
                    .filter(TIMESTAMP_INT.ge(start).and(TIMESTAMP_INT.le(end)))
                    .collect(Collectors.toList())
                    .filterNot { it.text in toFilterOut }

                if (utterances.isEmpty()) {
                    logger.info("No utterances found for TOC entry: {}", toc.startDate)
                    return@mapNotNull null
                }

                val grouped = utterances.joinToString("\n") { "${it.speaker}: ${it.text}" }

                toc.apply {
                    concatenatedUtterances = grouped
                    this.utterances = utterances
                }
            }
            .onEach {
                tocDataRepository.save(it)
                logger.info("Grouped utterances for TOC entry: {}", it.startDate)
            }

        logger.info("Grouping complete")
    }

    fun getAllWithoutQuestions(): List<TOCData> {
        return tocDataRepository.findAll()
            .filter {
                !it.concatenatedUtterances.isNullOrBlank() &&
                        it.questions.isNullOrEmpty()
            }
    }

    fun getAllWithQuestions(): List<TOCData> {
        return tocDataRepository.findAll().filter { !it.questions.isNullOrEmpty() }
    }

    fun getAllWithoutSummary(): List<TOCData> {
        return tocDataRepository.findAll()
            .filter {
                !it.concatenatedUtterances.isNullOrBlank() &&
                        it.summary.isNullOrEmpty()
            }
    }

    fun getAllWithSummary(): List<TOCData> {
        return tocDataRepository.findAll().filter { !it.summary.isNullOrEmpty() }
    }

    fun saveAll(tocData: List<TOCData>): List<TOCData> {
        tocDataRepository
        return tocDataRepository.saveAll(tocData)
    }

    fun updateSummary(tocData: List<TOCData>) {
        tocData.forEach { toc ->
            tocDataRepository.updateField(toc, `TOCData$`.SUMMARY, toc.summary)
        }
    }

    fun updateQuestions(tocData: List<TOCData>) {
        tocData.forEach { toc ->
            tocDataRepository.updateField(toc, `TOCData$`.QUESTIONS, toc.questions)
        }
    }
}